File: problems/38-advanced-graph-combinations/004-articulation-points/java/Solution.java
import java.util.*;
public class Solution {
 int time; int[] tin,low; boolean[] cut; List<Integer>[] g;
 public List<Integer> articulationPoints(int n,int[][] edges){g=new ArrayList[n];for(int i=0;i<n;i++)g[i]=new ArrayList<>();for(int[]e:edges){g[e[0]].add(e[1]);g[e[1]].add(e[0]);}tin=new int[n];low=new int[n];cut=new boolean[n];Arrays.fill(tin,-1);for(int i=0;i<n;i++)if(tin[i]<0)dfs(i,-1);List<Integer>a=new ArrayList<>();for(int i=0;i<n;i++)if(cut[i])a.add(i);return a;}
 void dfs(int v,int p){tin[v]=low[v]=time++;int children=0;for(int to:g[v]){if(to==p)continue;if(tin[to]>=0)low[v]=Math.min(low[v],tin[to]);else{dfs(to,v);low[v]=Math.min(low[v],low[to]);if(p!=-1&&low[to]>=tin[v])cut[v]=true;children++;}}if(p==-1&&children>1)cut[v]=true;}
}

File: problems/38-advanced-graph-combinations/004-articulation-points/python/solution.py
class Solution:
 def articulation_points(self,n,edges):
  g=[[] for _ in range(n)]
  for u,v in edges:g[u].append(v);g[v].append(u)
  tin=[-1]*n;low=[0]*n;cut=[False]*n;time=0
  def dfs(v,p):
   nonlocal time
   tin[v]=low[v]=time;time+=1;children=0
   for to in g[v]:
    if to==p:continue
    if tin[to]>=0:low[v]=min(low[v],tin[to])
    else:
     dfs(to,v);low[v]=min(low[v],low[to]);children+=1
     if p!=-1 and low[to]>=tin[v]:cut[v]=True
   if p==-1 and children>1:cut[v]=True
  for v in range(n):
   if tin[v]<0:dfs(v,-1)
  return [i for i,x in enumerate(cut) if x]
